File: ios/RegisterDropCounter/AppModel.swift
import Foundation
import SwiftUI
import Combine

@MainActor
final class AppModel: ObservableObject {
    @Published var base = 400
    @Published var registers: [Counts]
    @Published var names: [String]
    @Published var history: [HistoryEntry] = []
    @Published var active = 0
    @Published var bag = ""
    @Published var initials = ""

    static let version = "2.12.0"
    static let releaseDate = "2026-09-15"

    private let historyLimit = 200

    init() {
        registers = (0..<DropEngine.registerCount).map { _ in Counts() }
        names = (0..<DropEngine.registerCount).map { "R\($0 + 1)" }
        load()
        #if DEBUG
        DropEngineSelfTest.run()
        #endif
    }

    func result(_ i: Int) -> RegisterResult {
        DropEngine.compute(registers[i], baseDollars: base)
    }

    var results: [RegisterResult] {
        registers.map { DropEngine.compute($0, baseDollars: base) }
    }

    func setCount(register: Int, denom: Denom, value: Int) {
        registers[register][denom] = value
        persistLive()
    }

    func setName(register: Int, name: String) {
        let t = name.trimmingCharacters(in: .whitespacesAndNewlines)
        names[register] = t.isEmpty ? "R\(register + 1)" : String(t.prefix(20))
        persistNames()
    }

    func clearRegister(_ i: Int) {
        snapshot(kind: HistoryKind.register, index: i)
        registers[i] = Counts()
        persistLive()
    }

    func clearAll() {
        snapshot(kind: HistoryKind.all, index: nil)
        registers = (0..<DropEngine.registerCount).map { _ in Counts() }
        persistLive()
    }

    func undoClear() {
        guard let last = history.first else { return }
        if last.kind == HistoryKind.register, let i = last.registerIndex, i >= 0, i < last.registers.count {
            registers[i] = last.registers[i]
            active = i
        } else {
            base = last.base
            registers = last.registers
        }
        persistLive()
    }

    func restore(_ id: String) {
        guard let e = history.first(where: { $0.id == id }) else { return }
        base = e.base
        registers = e.registers
        persistLive()
    }

    func clearHistory() {
        history = []
        persistHistory()
    }

    func loadSample() {
        base = 400
        registers = SampleData.registers
        persistLive()
    }

    private func snapshot(kind: String, index: Int?) {
        if kind == HistoryKind.register {
            guard let i = index, registers[i].hasCount else { return }
        } else if !registers.contains(where: { $0.hasCount }) {
            return
        }
        let e = HistoryEntry(
            id: UUID().uuidString,
            at: Date().timeIntervalSince1970,
            kind: kind,
            registerIndex: kind == HistoryKind.register ? index : nil,
            base: base,
            registers: registers
        )
        history.insert(e, at: 0)
        if history.count > historyLimit { history = Array(history.prefix(historyLimit)) }
        persistHistory()
    }

    private func docs() -> URL {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
    }

    private func persistLive() {
        let payload: [String: Any] = [
            "base": base,
            "registers": registers.map(\.n),
        ]
        UserDefaults.standard.set(payload, forKey: "rdc.live")
        try? JSONSerialization.data(withJSONObject: payload).write(to: docs().appendingPathComponent("RegisterDropCounter.state.json"))
    }

    private func persistNames() {
        UserDefaults.standard.set(names, forKey: "rdc.names")
    }

    private func persistHistory() {
        if let data = try? JSONEncoder().encode(history) {
            UserDefaults.standard.set(data, forKey: "rdc.history")
            try? data.write(to: docs().appendingPathComponent("RegisterDropCounter.history.json"))
        }
    }

    private func persistSlip() {
        UserDefaults.standard.set(bag, forKey: "rdc.bag")
        UserDefaults.standard.set(initials, forKey: "rdc.initials")
    }

    func saveSlipFields() { persistSlip() }

    private func load() {
        if let n = UserDefaults.standard.array(forKey: "rdc.names") as? [String] {
            names = (0..<DropEngine.registerCount).map { i in
                let t = i < n.count ? n[i].trimmingCharacters(in: .whitespaces) : ""
                return t.isEmpty ? "R\(i + 1)" : String(t.prefix(20))
            }
        }
        if let live = UserDefaults.standard.dictionary(forKey: "rdc.live") {
            if let b = live["base"] as? Int, DropEngine.baseOptions.contains(b) { base = b }
            if let rows = live["registers"] as? [[Int]] {
                registers = (0..<DropEngine.registerCount).map { i in
                    Counts(n: i < rows.count ? rows[i] : [])
                }
            }
        }
        if let data = UserDefaults.standard.data(forKey: "rdc.history"),
           let h = try? JSONDecoder().decode([HistoryEntry].self, from: data) {
            history = h
        }
        bag = UserDefaults.standard.string(forKey: "rdc.bag") ?? ""
        initials = UserDefaults.standard.string(forKey: "rdc.initials") ?? ""
    }
}

enum SampleData {
    static var registers: [Counts] {
        var out = (0..<DropEngine.registerCount).map { _ in Counts() }
        out[0] = Counts(n: [0, 6, 48, 27, 0, 0, 0, 0, 109, 0, 5, 10, 103, 6, 12])
        out[1] = Counts(n: [0, 51, 18, 16, 0, 0, 0, 0, 100, 0, 2, 1, 98, 3, 5])
        out[2] = Counts(n: [5, 63, 40, 0, 0, 0, 0, 0, 89, 0, 19, 25, 154, 2, 13])
        return out
    }
}

enum DropEngineSelfTest {
    static func run() {
        var c = Counts()
        c[.nickel] = 6; c[.dime] = 48; c[.quarter] = 27
        c[.one] = 109; c[.five] = 5; c[.ten] = 10
        c[.twenty] = 103; c[.fifty] = 6; c[.hundred] = 12
        let r = DropEngine.compute(c, baseDollars: 400)
        assert(r.amountCents == 380585)
        assert(r.dropCents == 340585)
        assert(r.leftCents == 40000)
        assert(r.balanced)
        assert(r.drop[.twenty] == 95)
        assert(r.drop[.hundred] == 12)
        assert(r.left[.one] == 109)
    }
}

File: android/app/src/main/java/com/ronaldrobbins/registerdropcounter/AppModel.kt
package com.ronaldrobbins.registerdropcounter

import android.app.Application
import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

class AppModel(app: Application) : AndroidViewModel(app) {
    var base by mutableIntStateOf(400)
        private set
    var active by mutableIntStateOf(0)
        private set
    var bag by mutableStateOf("")
    var initials by mutableStateOf("")

    val registers = mutableStateListOf<Counts>().apply {
        repeat(DropEngine.REGISTER_COUNT) { add(Counts()) }
    }
    val names = mutableStateListOf<String>().apply {
        repeat(DropEngine.REGISTER_COUNT) { add("R${it + 1}") }
    }
    val history = mutableStateListOf<HistoryEntry>()

    init {
        load()
        DropEngineSelfTest.run()
    }

    fun result(i: Int): RegisterResult = DropEngine.compute(registers[i], base)

    fun setActive(i: Int) {
        active = i.coerceIn(0, DropEngine.REGISTER_COUNT - 1)
    }

    fun setBase(v: Int) {
        if (v in DropEngine.baseOptions) {
            base = v
            persistLive()
        }
    }

    fun setCount(register: Int, denom: Denom, value: Int) {
        val next = registers[register].copyOf()
        next[denom] = value
        registers[register] = next
        persistLive()
    }

    fun setName(register: Int, name: String) {
        val t = name.trim().replace(Regex("\\s+"), " ").take(20)
        names[register] = t.ifEmpty { "R${register + 1}" }
        persistNames()
    }

    fun clearRegister(i: Int) {
        snapshot(HistoryEntry.REGISTER, i)
        registers[i] = Counts()
        persistLive()
    }

    fun clearAll() {
        snapshot(HistoryEntry.ALL, null)
        for (i in 0 until DropEngine.REGISTER_COUNT) registers[i] = Counts()
        persistLive()
    }

    fun undoClear(): Boolean {
        val last = history.firstOrNull() ?: return false
        if (last.kind == HistoryEntry.REGISTER) {
            val i = last.registerIndex ?: return false
            if (i in last.registers.indices) {
                registers[i] = last.registers[i].copyOf()
                active = i
            }
        } else {
            base = last.base
            for (i in 0 until DropEngine.REGISTER_COUNT) {
                registers[i] = last.registers.getOrElse(i) { Counts() }.copyOf()
            }
        }
        persistLive()
        return true
    }

    fun restore(id: String): Boolean {
        val e = history.firstOrNull { it.id == id } ?: return false
        base = e.base
        for (i in 0 until DropEngine.REGISTER_COUNT) {
            registers[i] = e.registers.getOrElse(i) { Counts() }.copyOf()
        }
        persistLive()
        return true
    }

    fun clearHistory() {
        history.clear()
        persistHistory()
    }

    fun loadSample() {
        base = 400
        val sample = SampleData.registers()
        for (i in 0 until DropEngine.REGISTER_COUNT) registers[i] = sample[i]
        persistLive()
    }

    fun saveSlipFields() {
        prefs().edit().putString("bag", bag).putString("initials", initials).apply()
    }

    private fun snapshot(kind: String, index: Int?) {
        if (kind == HistoryEntry.REGISTER) {
            val i = index ?: return
            if (!registers[i].hasCount) return
        } else if (registers.none { it.hasCount }) {
            return
        }
        val e = HistoryEntry(
            id = UUID.randomUUID().toString(),
            at = System.currentTimeMillis(),
            kind = kind,
            registerIndex = if (kind == HistoryEntry.REGISTER) index else null,
            base = base,
            registers = registers.map { it.copyOf() },
        )
        history.add(0, e)
        while (history.size > 200) history.removeAt(history.lastIndex)
        persistHistory()
    }

    private fun prefs() = getApplication<Application>().getSharedPreferences("rdc", Context.MODE_PRIVATE)

    private fun persistLive() {
        val arr = JSONArray()
        registers.forEach { c ->
            val row = JSONArray()
            c.n.forEach { row.put(it) }
            arr.put(row)
        }
        prefs().edit().putInt("base", base).putString("registers", arr.toString()).apply()
    }

    private fun persistNames() {
        prefs().edit().putString("names", JSONArray(names.toList()).toString()).apply()
    }

    private fun persistHistory() {
        val arr = JSONArray()
        history.forEach { e ->
            val o = JSONObject()
            o.put("id", e.id)
            o.put("at", e.at)
            o.put("kind", e.kind)
            o.put("registerIndex", e.registerIndex ?: -1)
            o.put("base", e.base)
            val regs = JSONArray()
            e.registers.forEach { c ->
                val row = JSONArray()
                c.n.forEach { row.put(it) }
                regs.put(row)
            }
            o.put("registers", regs)
            arr.put(o)
        }
        prefs().edit().putString("history", arr.toString()).apply()
    }

    private fun load() {
        val p = prefs()
        val b = p.getInt("base", 400)
        if (b in DropEngine.baseOptions) base = b
        p.getString("registers", null)?.let { raw ->
            val arr = JSONArray(raw)
            for (i in 0 until DropEngine.REGISTER_COUNT) {
                val row = if (i < arr.length()) arr.getJSONArray(i) else JSONArray()
                val vals = List(Denom.entries.size) { di -> if (di < row.length()) row.optInt(di) else 0 }
                registers[i] = Counts.fromList(vals)
            }
        }
        p.getString("names", null)?.let { raw ->
            val arr = JSONArray(raw)
            for (i in 0 until DropEngine.REGISTER_COUNT) {
                val t = if (i < arr.length()) arr.optString(i).trim() else ""
                names[i] = t.ifEmpty { "R${i + 1}" }.take(20)
            }
        }
        p.getString("history", null)?.let { raw ->
            val arr = JSONArray(raw)
            history.clear()
            for (i in 0 until arr.length()) {
                val o = arr.getJSONObject(i)
                val regsArr = o.optJSONArray("registers") ?: JSONArray()
                val regs = MutableList(DropEngine.REGISTER_COUNT) { Counts() }
                for (r in 0 until DropEngine.REGISTER_COUNT) {
                    val row = if (r < regsArr.length()) regsArr.getJSONArray(r) else JSONArray()
                    val vals = List(Denom.entries.size) { di -> if (di < row.length()) row.optInt(di) else 0 }
                    regs[r] = Counts.fromList(vals)
                }
                val idx = o.optInt("registerIndex", -1)
                history.add(
                    HistoryEntry(
                        id = o.optString("id"),
                        at = o.optLong("at"),
                        kind = o.optString("kind"),
                        registerIndex = if (idx >= 0) idx else null,
                        base = o.optInt("base", 400),
                        registers = regs,
                    ),
                )
            }
        }
        bag = p.getString("bag", "") ?: ""
        initials = p.getString("initials", "") ?: ""
    }
}
